Tools/AP_Periph/esc_apd_telem.cpp

```diff
@@ -0,0 +1,103 @@
+/*
+  ESC Telemetry for Hobbywing Pro 80A HV ESC. This will be
+  incorporated into a broader ESC telemetry library in ArduPilot
+  master in the future
+
+  This protocol only allows for one ESC per UART RX line, so using a
+  CAN node per ESC works well.
+ */
+#include "esc_apd_telem.h"
+#include <AP_HAL/utility/sparse-endian.h>
+#include <AP_Math/crc.h>
+#include <AP_Math/definitions.h>
+#include <string.h>
+
+#ifdef HAL_PERIPH_ENABLE_ESC_APD
+
+extern const AP_HAL::HAL& hal;
+
+#define TELEM_HEADER 0x9B
+#define TELEM_LEN    0x16
+
+ESC_APD_Telem::ESC_APD_Telem(void) { }
+
+void ESC_APD_Telem::init(AP_HAL::UARTDriver *_uart, float num_poles) {
+    uart = _uart;
+    uart->begin(115200);
+
+    pole_count = num_poles;
+}
+
+bool ESC_APD_Telem::update() {
+    uint32_t n = uart->available();
+    if (n == 0) {
+        return false;
+    }
+
+    // don't read too much in one loop to prevent too high CPU load
+    if (n > 500) {
+        n = 500;
+        uart->discard_input();
+    }
+
+    bool ret = false;
+
+    while (n--) {
+        uint8_t b = uart->read();
```

Comment:
this whole thing is best handled with a circular buffer where you only do any work if the current and previous are 0xFF. It can be done easy without looking/looping the buffer history if you compare the current buffer byte /w the new read() value. If both are 0xFF you load it and compute, else load and continue;

Reply:
I'm not sure that's actually really true here. I'd still have to pop everything back out of the buffer again. If I lost sync then the existing memmove is worse, however on a well behaved link (which is the general case) a circular buffer means we have to copy all the data at least one extra time then we would before. Given that we should normally be synced I'm inclined to prefer the current structure instead.